src/theory/sets/theory_sets_private.cpp

```diff
@@ -1978,50 +1020,41 @@ bool TheorySetsPrivate::collectModelInfo(TheoryModel* m)
     return false;
   }
 
+  NodeManager* nm = NodeManager::currentNM();
   std::map< Node, Node > mvals;
-  for( int i=(int)(d_set_eqc.size()-1); i>=0; i-- ){
-    Node eqc = d_set_eqc[i];
+  // If cardinality is enabled, we need to use the ordered equivalence class
+  // list computed by the cardinality solver, where sets equivalence classes
+  // are assigned model values based on their position in the cardinality
+  // graph.
+  const std::vector<Node>& sec = d_card_enabled
+                                     ? d_cardSolver->getOrderedSetsEqClasses()
+                                     : d_state.getSetsEqClasses();
+  Valuation& val = getValuation();
+  for (int i = (int)(sec.size() - 1); i >= 0; i--)
+  {
+    Node eqc = sec[i];
     if( termSet.find( eqc )==termSet.end() ){
       Trace("sets-model") << "* Do not assign value for " << eqc << " since is not relevant." << std::endl;
     }else{
       std::vector< Node > els;
-      bool is_base = !d_card_enabled || ( d_nf[eqc].size()==1 && d_nf[eqc][0]==eqc );
+      bool is_base = !d_card_enabled || d_cardSolver->isModelValueBasic(eqc);
       if( is_base ){
         Trace("sets-model") << "Collect elements of base eqc " << eqc << std::endl;
         // members that must be in eqc
-        std::map< Node, std::map< Node, Node > >::iterator itm = d_pol_mems[0].find( eqc );
-        if( itm!=d_pol_mems[0].end() ){
-          for( std::map< Node, Node >::iterator itmm = itm->second.begin(); itmm != itm->second.end(); ++itmm ){
-            Node t = NodeManager::currentNM()->mkNode( kind::SINGLETON, itmm->first );
+        const std::map<Node, Node>& emems = d_state.getMembers(eqc);
+        if (!emems.empty())
+        {
+          for (const std::pair<const Node, Node>& itmm : emems)
+          {
+            Node t = nm->mkNode(kind::SINGLETON, itmm.first);
```

Comment:
Suggestion: Why don't we construct sets using `insert` instead of `union` which is more verbose? compare this
```
(define-fun S () (Set Int) (insert 1 2 3 (singleton 4)))
```
versus
```
(define-fun S () (Set Int) (union (union (union (singleton 1) (singleton 2)) (singleton 3)) (singleton 4)))
```

Reply:
`insert` is syntax sugar. If we insisted that the rewritten form involved `insert`, then our decision procedures would have to handle `union` and `insert`, thus making the code for handling set union twice as complicated.